Development

CihatAltiparmak opened the pull request.

CihatAltiparmak committed:
Added config files

CihatAltiparmak committed:
Beginning of Middleware Benchmark Implementation

CihatAltiparmak committed:
Refactor benchmark code

- Refactored test_scenario_perception_pipeline.cpp
- Added event handler because of interaction problem between panda_arm_spawner and test_scenario_perception_pipeline node

CihatAltiparmak committed:
Integrated ros-industrial-ci and updated some external dependencies

CihatAltiparmak committed:
Changed the function namings by camelCase and some changes in launch files

CihatAltiparmak committed:
Made test case creation flexible and removed unnecessary imports from launch file

CihatAltiparmak committed:
Added how to run and how to install sections to README and some pre-fix

CihatAltiparmak committed:
Fix CI for UPSTREAM_WORKSPACE

- Added upstream workspace envrionment to CI
- Changed ROS version in CI
- Fixed CMakeList.txt and package.xml based on the feedbacks of CI
- Fixed upstream repository urls
- Changed test cases for more convenient benchmarking

CihatAltiparmak committed:
Rearranged directory tree structure and some modifications

- Added middleware implementations to exec_depend ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647678198 )
- Refactored package directory tree structure ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647697922 )
- Added namespace ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647703131 )

CihatAltiparmak committed:
Added documentation and fixed package.xml

- Added documentation for each methods and variables except global variables ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647698884 )
- Put constants into anoymous namespace ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647701794 )
- Modified condition checking in senfTargetPose ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647681828 )
- Fixed middleware packages in package.xml
- Renamed the test case config

CihatAltiparmak committed:
Removed ScenarioPerceptionPipelineTestCaseGenerator class

- Removed ScenarioPerceptionPipelineTestCaseGenerator class and added its methods to ScenarioPerceptionPipelineFixture
- Renamed ScenarioPerceptionPipelineFixture/selectTestCase method as ScenarioPerceptionPipelineFixture/selectTestCase
- Added documentation for ScenarioPerceptionPipelineFixture/setUp and ScenarioPerceptionPipelineFixture/tearDown methods

CihatAltiparmak committed:
Renamed the files related to scenario_percpetion_pipeline

CihatAltiparmak committed:
Merge pull request #2 from CihatAltiparmak/fix/refactor_codebase

Creating Repository Structure

CihatAltiparmak committed:
Feature/add documents (#4)

* Created documentations

* Revised documentation

CihatAltiparmak commented:
Hello @sjahr  and @henningkayser , could you review this PR? I need your help for this stage. I have added a documentation. I have already answered and applied necessary modifications for @sjahr 's review in #1

sjahr commented on CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+                      PUBLIC "benchmark::benchmark" ${YAML_CPP_LIBRARIES})
+
+install(TARGETS scenario_perception_pipeline_benchmark_main
+        DESTINATION lib/${PROJECT_NAME})
```
Replace ${PROJECT_NAME}  with moveit_middleware_benchmark (also applies to comment below)

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/0ba81e8b859e5ef794b13d768238d1c2cce5b58f commit

sjahr commented on README.md:
````diff
@@ -1 +1,33 @@
+
+```
+
+It will be created the json file for benchmarking results after finishing benchmark code execution.
````
Maybe you can add a small section here how to visualize the results (You can also open an issue and do that in a follow-up PR)

CihatAltiparmak replied:
I actually implemented visualizer script in python. It would be great to prepare visualization relavant stuff in follow-up PR. But if you wonder it, i run benchmarks separately per each middleware (rmw_zenoh, rmw_cyclonedds and rmw_fastrtps). Then i move this files to where the script is located and run scripts. Box plot is created magically after running script. Script is [here](https://gist.github.com/CihatAltiparmak/899090aca71223fbb964569f83265a49).

sjahr replied:
Ok :+1: I think you can add it to this repository in a folder called scripts.

sjahr commented on config/scenario_perception_pipeline_test_cases.yaml:
```diff
@@ -0,0 +1,157 @@
+test_cases:
```
Can you add a short comment that describes the test cases and why you selected these ones? It can also be something like `X randomly selected poses for the perception pipeline benchmark` if there is nothing special about them

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/3857b680b32a083d9d95825405ecf06e9973e0a8 commit

sjahr commented on README.md:
```diff
@@ -1 +1,33 @@
\ No newline at end of file
+# moveit_middleware_benchmark
+
+Under Construction
```
Add a small descriptive abstract (3-4 sentences about the repositories components and their purpose)
```
This repository contains ...
The perception pipeline benchmark can be used to measure ...
...
```

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/fa4710c5de5217be352e51797c365f8fc1e1ae1f commit

sjahr commented on docs/how_to_install.md:
```diff
@@ -0,0 +1,25 @@
+cd ws/src
+
+# clone this repository
+git clone git@github.com:CihatAltiparmak/moveit_middleware_benchmark.git -b fix/refactor_codebase
```
Is that branch still the one to use?

CihatAltiparmak replied:
No, this one was for under construction period. Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/88543a0213a9b9d3add4116ee954a7ce5518d05c commit. But it is better to work on development at the moment. Because we still haven't merged this branch(development) to main.

sjahr commented on docs/how_to_run.md:
```diff
@@ -0,0 +1,9 @@
+
+### [Perception Pipeline Benchmark](scenarios/perception_pipeline_benchmark.md)
+
+This benchmark measures the elapsed time by which the determined path is sent for the robot to follow. This benchmark calculates `elapsed time`, `success_number` and `failure_number`. `elapsed_time` is used for measuring how much time the pipeline takes. `success_number` is used for presenting successfully plannings and `failure_number` is used for presenting failed plannings.
```
How to you differentiate between time required by the middleware communication layer  vs the planner/ execution algorithms?

CihatAltiparmak replied:
The elapsed time means real time in benchmark results created by google benchmark tool. It seems there is no other way to measure the time plan-execution jobs took and the time data communication jobs took. Because i just implemented benchmark in the client side (move_group_interface). Instead, i follow the strategy that 

* to measure elapsed time (wall time, real time in other meaning)  
  * send pose msg for the robot to plan
  * wait for servers response (data communication jobs + planning jobs) 
  * send plan msg for robot to execute trajectory (data communication jobs + execution jobs) 
* to do this measurements per each middleware
* compare the results

If we make the total time of planning jobs and execution jobs same for each middleware, planning jobs and execution jobs don't affect comparison of middleware. To send much goal pose (i'm speaking for perception pipeline) decreases randomness of ompl planner etc. The real aim isn't to mesaure the time, but to compare the time per each middleware.

sjahr commented on docs/how_to_run.md:
```diff
@@ -0,0 +1,9 @@
+
+This benchmark measures the elapsed time by which the determined path is sent for the robot to follow. This benchmark calculates `elapsed time`, `success_number` and `failure_number`. `elapsed_time` is used for measuring how much time the pipeline takes. `success_number` is used for presenting successfully plannings and `failure_number` is used for presenting failed plannings.
+
+Firstly, `node` and `move_group_interface`in SetUp are created before each benchmark. `poses` inside `nav_msgs/msg/Path` is sent one by one to plan trajectory for robot. If planning is failed, only `failure_rate` is increased. If planning is successful, the trajectory_plan which move_group_server plan is sent via `move_group_interface` to start the execution of this planned trajectory. Then `succ ...
```
Does the selected middleware have any impact on the success / failure of the planning request? I'd assume it shouldn't and in that case these numbers might be not super relevant for the benchmark. What do you think?

CihatAltiparmak replied:
You're right. But we are working on experimental middleware as well. I'm trying to follow each updates in rmw_zenoh. It is possible that forward updates on rmw_zenoh break the benchmark pipeline. rmw_zenoh had problem for multithreaded executor before(It seems it's fixed now but there is still no guarantee that rmw_zenoh will not break our benchmark) and this bug sometimes broke our scenario before. (communication between action_client and action_server). To sum up, i want not to remove this part because of above reasons. But if you think your idea is still better than mine, i can change this part.

sjahr commented on docs/index.md:
```diff
@@ -0,0 +1,6 @@
+# MoveIt Middleware Benchmark
+
+This is a package created for measuring middleware effects on various scenarios. This package contains some scenarios such as perception pipeline.
```
I like that! Move it directly into the README.md file. The tutorials use and index.md file because it is used to build the website but for this repos, I think it is not necessary

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/57a5a2ce4de5ae5663a36a34fb412d9509c7e6d6 commit. I hope this one improves README.md. Btw i also added CI badge :smiley:

sjahr commented on docs/scenarios/perception_pipeline_benchmark.md:
```diff
@@ -0,0 +1,193 @@
+
+## How to create test cases
+
+You can add your test cases to `scenario_perception_pipeline_test_cases.yaml` file. For benchmarking of the scenario of perception pipeline, your test case must be presented in a type of nav_msgs/msg/Path. `poses` section presents `poses` field of `nav_msgs/msg/Path` type. There is a following example test case file.
```
How relevant is it which pose I select for the middleware benchmark?

CihatAltiparmak replied:
Not to select pose that planner gets hard to calculate path is better for perception pipeline. Because we are measuring for middleware, not for planner.  Should i add this part to documentation, @sjahr ?

sjahr replied:
Yes, I think that would be nice. But you can also do that in a follow-up PR

CihatAltiparmak replied:
Let's do this one as resolved. Maybe we can take a look at it later together with @henningkayser

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+
+#include <rclcpp/rclcpp.hpp>
+#include <benchmark/benchmark.h>
+#include <moveit/move_group_interface/move_group_interface.h>
```
Nitpick: You could group all moveit headers together

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/066e8dff0e01e2824773f62826e58375a38fc4d4 commit

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+  bool sendTargetPose(const geometry_msgs::msg::Pose& target_pose);
+
+private:
+  /** move_group_interface_ptr to communicate with move_grouo_server */
```
```suggestion
  /** move_group_interface_ptr to communicate with move_group_server */
```

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+class ScenarioPerceptionPipelineFixture : public benchmark::Fixture
+{
+protected:
+  /* ros node shared ptr*/
```
You can drop the comments that don't give more context like this

CihatAltiparmak replied:
Done it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+
+class ScenarioPerceptionPipelineFixture : public benchmark::Fixture
+{
+protected:
```
Nitpick: I'd put the protected class members after the public ones

CihatAltiparmak replied:
Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+   */
+  void createTestCases();
+
+  /** \brief Given a \e test_case_index, returns relevant test case.
```
What's the selection criteria?

CihatAltiparmak replied:
I'm passing parameters to benchmark node like below. User selects one of the test cases and only selected test case is benchmarked.

https://github.com/CihatAltiparmak/moveit_middleware_benchmark/blob/caa54960dd6febab1c82f7b63b93a38616339cda/launch/scenario_perception_pipeline_benchmark.launch.py#L128-L134

sjahr commented on include/moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp:
```diff
@@ -0,0 +1,163 @@
+  /** \brief Given a \e yaml_file_name, reads test cases from given yaml file
+   *  and adds this read test cases to \e test_cases_
+   *  \param [in] yaml_file_name the path of yaml file which includes test cases
+   *  \return nothing is returned
```
You can drop the `\return` comment for void functions

CihatAltiparmak replied:
Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/916a595a895dfc8febc655658a490f4051c8efa4 commit

sjahr commented on launch/scenario_perception_pipeline_benchmark.launch.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def generate_launch_description():
+    db_arg = DeclareLaunchArgument(
```
Do you need the db_arg?

CihatAltiparmak replied:
Removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

sjahr commented on launch/scenario_perception_pipeline_benchmark.launch.py:
```diff
@@ -0,0 +1,180 @@
+        .planning_scene_monitor(
+            publish_robot_description=True, publish_robot_description_semantic=True
+        )
+        .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
```
Is gripper controller,yaml the right controller config file?

CihatAltiparmak replied:
Honestly, i don't know, I have stolen most of it from moveit_resources repo :smile_cat: .

https://github.com/moveit/moveit_resources/blob/55d3ce0b8b95ddd224c3892ce49c37c0dc9c9361/panda_moveit_config/launch/demo.launch.py#L41-L49

sjahr commented on launch/scenario_perception_pipeline_benchmark.launch.py:
```diff
@@ -0,0 +1,180 @@
+    )
+
+    # Warehouse mongodb server
+    db_config = LaunchConfiguration("db")
```
I think this can be removed, it is not used, an Mongo DB might not even be available in ROS 2 (this information might be  a couple of months old though :sweat_smile: )

CihatAltiparmak replied:
Also removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

sjahr commented on launch/scenario_perception_pipeline_benchmark.launch.py:
```diff
@@ -0,0 +1,180 @@
+            joint_state_broadcaster_spawner,
+            panda_arm_controller_spawner,
+            panda_hand_controller_spawner,
+            mongodb_server_node,
```
Remove?

CihatAltiparmak replied:
Also removed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/caa54960dd6febab1c82f7b63b93a38616339cda commit

sjahr commented on package.xml:
```diff
@@ -0,0 +1,30 @@
+<package format="3">
+  <name>moveit_middleware_benchmark</name>
+  <version>0.0.0</version>
+  <description>Middleware Benchmark Tool For MoveIt Pipelines</description>
```
```suggestion
  <description>Middleware Benchmark Tool For MoveIt 2</description>
```

CihatAltiparmak replied:
Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/fa4710c5de5217be352e51797c365f8fc1e1ae1f commit

sjahr commented on src/scenario_perception_pipeline_benchmark_main.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "moveit_middleware_benchmark/scenarios/scenario_perception_pipeline.hpp"
```
Add license

CihatAltiparmak replied:
Fixed in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/dafc402a5f4ed41de6159a366e71677f5feea573 commit

sjahr commented on src/scenarios/scenario_perception_pipeline.cpp:
```diff
@@ -0,0 +1,156 @@
+  readTestCasesFromFile(TEST_CASES_YAML_FILE);
+}
+
+nav_msgs::msg::Path ScenarioPerceptionPipelineFixture::selectTestCase(size_t test_case_index)
```
In cases where the input is not changed you should make it const

CihatAltiparmak replied:
I fixed it in https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/1acebf40d7052970148dc6bd4a9ab84a2e5fd1bb commit. But I'm wondering  whether below code part should be returned const or not. Do you have any idea @sjahr ?

https://github.com/CihatAltiparmak/moveit_middleware_benchmark/blob/caa54960dd6febab1c82f7b63b93a38616339cda/src/scenarios/scenario_perception_pipeline.cpp#L131-L139

CihatAltiparmak replied:
I mark it as solved then.

sjahr commented on src/scenarios/scenario_perception_pipeline.cpp:
```diff
@@ -0,0 +1,156 @@
+void ScenarioPerceptionPipelineFixture::readTestCasesFromFile(const std::string& yaml_file_name)
+{
+  YAML::Node config = YAML::LoadFile(yaml_file_name.c_str());
+  for (YAML::const_iterator it = config["test_cases"].begin(); it != config["test_cases"].end(); ++it)
```
When possible, prefere `.at(i)` over `[i]`, while it is a bit slower, you get a nicer segfault message when trying to access something out of bounds

CihatAltiparmak replied:
I couldn't understand where i apply this changes. @sjahr Could you explain this part in more detailed ? Btw i try to use your suggestion everywhere.

sjahr replied:
Oh you're right, I guess I thought that config is a vector but I see that the type is `YAML::Node`.

CihatAltiparmak replied:
Let's do this one as resolved. Maybe we can take a look at it later together with @henningkayser

sjahr commented on src/scenarios/scenario_perception_pipeline.cpp:
```diff
@@ -0,0 +1,156 @@
+    auto sc = ScenarioPerceptionPipeline(move_group_interface_ptr_);
+    auto [success_number, failure_number] = sc.runTestCase(selected_test_case);
+    st.counters["success_number"] = success_number;
+    st.counters["failure_number"] = failure_number;
```
I am confused, where is the elapsed time measured?

CihatAltiparmak replied:
Google benchmark library handles this stuff automatically. `st.counters` is used optionally for additional outputs like failure rate, sucess_rate. For more, you can take a cursory look at https://github.com/google/benchmark/blob/main/docs/user_guide.md#counter-reporting If you are still confused, ping me again.

sjahr replied:
Thanks!

sjahr reviewed (commented):
Well-structured code and good amount of in-code documentation, thanks!
 Is this PR supposed to measure the elapsed time as will or is it just setting up the basic infrastructure? Also I don't understand how you are differentiating between the middleware processing time and the planning time (for the middleware benchmark we only care for the first).

CihatAltiparmak committed:
Renamed ${PROJECT_NAME} as moveit_middleware_benchmark in CmakeLists.txt See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663690770 )

CihatAltiparmak committed:
Added comments to make the example test cases more understandable. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663708224 )

CihatAltiparmak committed:
Modified README.md and fixed typo of description in package.xml

- Added some explanation. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663711001 )
- Fixed typo of description in package.xml. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663751883 )

CihatAltiparmak committed:
Fixed how_to_install documentation. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663712506 )

CihatAltiparmak committed:
Deleted index.md. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663731938 )

CihatAltiparmak committed:
Grouped all MoveIt headers into one place. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663735212 )

CihatAltiparmak committed:
Put protected members after public ones and removed \return tag from void function's documentation.

See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663739560 )
See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663743011 )

CihatAltiparmak committed:
Made the inputs which is unnecessary to change constant. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663755672 )

CihatAltiparmak committed:
Add LICENSE to src/scenario_perception_pipeline_benchmark_main.cpp. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663753425 )

CihatAltiparmak committed:
Remove mongo_db_warehouse stuff in launch/scenario_perception_pipeline_benchmark.launch.py

- See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663749042 )
- See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1663744156 )

CihatAltiparmak commented:
> Well-structured code and good amount of in-code documentation, thanks! Is this PR supposed to measure the elapsed time as will or is it just setting up the basic infrastructure?

This PR creates a json report after running launch file s indicated in documentation. But launch doesn't close itself as soon as the execution of benchmark_node finishes. You must close it. For example report:

middleware_benchmark_results.json
```
{
  "context": {
    "date": "2024-06-26T17:06:43+03:00",
    "host_name": "jarbay51",
    "executable": "/home/cihat/ws_moveit/install/moveit_middleware_benchmark/lib/moveit_middleware_benchmark/benchmark_main",
    "num_cpus": 8,
    "mhz_per_cpu": 3500,
    "cpu_scaling_enabled": true,
    "caches": [
      {
        "type": "Data",
        "level": 1,
        "size": 32768,
        "num_sharing": 2
      },
      {
        "type": "Instruction",
        "level": 1,
        "size": 32768,
        "num_sharing": 2
      },
      {
        "type": "Unified",
        "level": 2,
        "size": 262144,
        "num_sharing": 2
      },
      {
        "type": "Unified",
        "level": 3,
        "size": 6291456,
        "num_sharing": 8
      }
    ],
    "load_avg": [0.614746,0.461914,0.591309],
    "library_build_type": "release"
  },
  "benchmarks": [
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 0,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.8798042440000244e+10,
      "cpu_time": 7.6061629999999985e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 1,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9686648539000092e+10,
      "cpu_time": 7.1083848000000000e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 2,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9630923466000244e+10,
      "cpu_time": 7.1635957000000030e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 3,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9480768437999359e+10,
      "cpu_time": 7.2967000000000000e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 4,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9488600278999908e+10,
      "cpu_time": 7.2488129999999985e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 5,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9655598063999603e+10,
      "cpu_time": 7.1309232000000060e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 6,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9581028653999969e+10,
      "cpu_time": 7.3491184999999955e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 7,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9675931750000015e+10,
      "cpu_time": 7.2911115000000030e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 8,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9572710986999481e+10,
      "cpu_time": 7.2098062000000015e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 9,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9543964225999847e+10,
      "cpu_time": 7.1122786000000000e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 10,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9449596418000504e+10,
      "cpu_time": 7.1320349999999925e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "iteration",
      "repetitions": 12,
      "repetition_index": 11,
      "threads": 1,
      "iterations": 1,
      "real_time": 8.9813698392999863e+10,
      "cpu_time": 7.3313866999999970e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time_mean",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "aggregate",
      "repetitions": 12,
      "threads": 1,
      "aggregate_name": "mean",
      "aggregate_unit": "time",
      "iterations": 12,
      "real_time": 8.9531459304499924e+10,
      "cpu_time": 7.2483596833333328e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time_median",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "aggregate",
      "repetitions": 12,
      "threads": 1,
      "aggregate_name": "median",
      "aggregate_unit": "time",
      "iterations": 12,
      "real_time": 8.9576869820499725e+10,
      "cpu_time": 7.2293096000000000e+07,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 2.0000000000000000e+01
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time_stddev",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "aggregate",
      "repetitions": 12,
      "threads": 1,
      "aggregate_name": "stddev",
      "aggregate_unit": "time",
      "iterations": 12,
      "real_time": 2.5284596896927872e+08,
      "cpu_time": 1.4249045911069282e+06,
      "time_unit": "ns",
      "failure_number": 0.0000000000000000e+00,
      "success_number": 0.0000000000000000e+00
    },
    {
      "name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time_cv",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ScenarioPerceptionPipelineFixture/test_scenario_perception_pipeline/process_time/real_time",
      "run_type": "aggregate",
      "repetitions": 12,
      "threads": 1,
      "aggregate_name": "cv",
      "aggregate_unit": "percentage",
      "iterations": 12,
      "real_time": 2.8241019517993102e-03,
      "cpu_time": 1.9658304131668747e-02,
      "time_unit": "ns",
      "failure_number": NaN,
      "success_number": 0.0000000000000000e+00
    }
  ]
}
```

CihatAltiparmak committed:
Fixed the issue not to be able to terminate all of nodes in launch as soon as benchmark node process is finished.

CihatAltiparmak commented:
Hello @sjahr , i think i'm ready for a second review. When you have any chance etc. to take a second look at this PR, could you review new commits? Btw i'm trying not to merge other branches to development branch so that you can review easier. Additionaly, I've created a github action for creating and pushing docker image from github registry. You will also have a chance to pull this docker image approximately in this morning :)

CihatAltiparmak committed:
Fixed ros2 node ptr destruction problem, which gives rise to finish with unsoundly

- First insights: It seems benhcmark nodes shouldn't store the shared ptrs of ros nodes etc.
  while rclcpp::shutdown is running
  Otherwise, benchmark sometimes finishes unsoundly like below.

	[ERROR] [1713850255.212319171] [bla_bla_node.rclcpp]: Error in destruction of rcl subscription handle: Failed to delete datareader, at ./src/subscription.cpp:52, at ./src/rcl/subscription.c:184
	cannot publish data, at ./src/rmw_publish.cpp:62 during '__function__'
	Fail in delete datareader, at ./src/rmw_service.cpp:104 during '__function__'

sjahr reviewed (dismissed):
@henningkayser Would you like to review it too before we merge it?

henningkayser commented on config/scenario_perception_pipeline_test_cases.yaml:
```diff
@@ -0,0 +1,171 @@
+          y: 0.5
+          z: 0.5
+        orientation:
+          w : 1.0
```
I think it's odd to just see the w component here. Can orientation be left empty altogether and default to the identity orientation?

CihatAltiparmak replied:
I couldn't undestand why you want to use orientation in default. I have run benchmark scenarios using this test case until now. I will find a more convenient test case for you. But it would be so great to firstly merge this development stuff and then apply the things in this review in follow-up PRs. Does it make sense @henningkayser ?

henningkayser replied:
I was only commenting on the readability. w=1 is the identity transform assuming it is a quaternion and the other components are set to 0. It would be easier to read this file if all components have to be specified (of course this would bloat a lot of redundant content), or to leave the orientation out and implicitly default to the identity orientation. I agree that this should not block merging the PR.

CihatAltiparmak replied:
I will add the other components in follow-up PR.

henningkayser commented on docs/how_to_install.md:
```diff
@@ -0,0 +1,25 @@
+## How To Install
+
+
+Firstly, setup the dependencies of moveit_middleware_benchmark repositories. It's suggested to test with rolling version of ROS.
```
Setup and build the `moveit_middleware_benchmark` including all source dependencies. This assumes that ROS Rolling is being used, but other ROS distros should work as well.

CihatAltiparmak replied:
Fixed it in this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/f2d0ab70c4d01f813484a58c99b6960b2c3206a7) Thanks

henningkayser commented on docs/scenarios/perception_pipeline_benchmark.md:
```diff
@@ -0,0 +1,208 @@
+## How To Run Perception Benchmark
+
+Firstly, source your ros version. It's suggested to test with rolling version of ROS.
```
just link to the install instructions to make sure everyone has followed these steps

CihatAltiparmak replied:
This commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/2b1c8b136302e678a25da2d74c329a8a8f3a63d4) should solve. Can you give a feedback again?

henningkayser commented on docs/scenarios/perception_pipeline_benchmark.md:
````diff
@@ -0,0 +1,208 @@
+```sh
+# go to your workspace
+cd ws
+source /opt/ros/rolling/setup.bash
````
why source rolling again? I would also assume that the workspace has been sourced initially as well

CihatAltiparmak replied:
You're right. I just forgot it. Fixed in this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/33677ffc956c0d8c273715485003501b8f69a4ff) Thanks.

henningkayser commented on docs/scenarios/perception_pipeline_benchmark.md:
```diff
@@ -0,0 +1,208 @@
+cd ws
+source /opt/ros/rolling/setup.bash
+source install/setup.bash
+export RMW_IMPLEMENTATION=rmw_zenoh_cpp # select your rmw_implementation to benchmark
```
Shouldn't this be set before initializing the zenoh router or any other ROS nodes?

CihatAltiparmak replied:
Fixed it this commit (https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/commits/33677ffc956c0d8c273715485003501b8f69a4ff) . Thanks.

CihatAltiparmak committed:
Fixed documentation of perception pipeline. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/files/398badab8d4c5e6ea80a6e489238b80d5e48423f#r1677879487 ) and ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5/files/398badab8d4c5e6ea80a6e489238b80d5e48423f#r1677880800 )

CihatAltiparmak committed:
Modifed explanations in how_to_install section of document. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1677876155 )

CihatAltiparmak committed:
Given a link which directs to How To Run page inside How To Install. See ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/5#discussion_r1677878485 )
